sequence_processing_pipeline/Pipeline.py

```diff
@@ -215,9 +236,9 @@ def add(self, job):
     def _validate_sample_sheet(self, sample_sheet_path):
         """
         Performs additional validation for sample-sheet on top of metapool.
-        :return: If successful, an empty list of strings and a valid
-                 sample-sheet. If unsuccessful, a list of warning and error
-                 messages and None.
+        :return: If successful, a valid sample-sheet. Raises descriptive
```

Comment:
The code that handles warnings/errors in this method is a little strange (noticed due to the change in the docstring). As far as I can tell, it adds all messages to the variable msg and then loops over it to add split between warnings/errors, is this really desirable or would it be better to keep them separate (and avoid looping multiple times)? Also, in line 281 the code raises a ErrorMessage vs a  PipelineError, is that expected?

Reply:
The reasoning behind this is that a Pipeline can't be instantiated with an invalid sample-sheet or mapping-file and the constructor raising an Error is preferable to returning an invalid object.

This is actually the part of the code that generates the error message seen by the user when a sample-sheet contains errors. For example, metapool may find a number of issues w/a sample-sheet that SPP wants to display to the user. Rather than return the first error and have the user repeatedly submit and experience errors, we aggregate the error messages and return them as a single message. This seems fair and the most truthful in that no single issue alone was the reason for the failure.

Warnings on the other hand are not aggregated as they don't prevent the sample-sheet/mapping-file from being valid. Those are stored in the pipeline as a list under self.warnings.

Yes, Line 281 needs to create an ErrorMessage that can later be aggregated with the other ErrorMessages from the KLSampleSheet() object as described above. Raising a PipelineError instead would hide the other messages.

Reply:
If warnings are not aggregated, why do we need the error check loop (line 248)? Wouldn't it be safe to assume they are all errors?

Reply:
msgs is going to be a list of both ErrorMessages and WarningMessages. Hence, we need to generate a separate list of ErrorMessages, both to identify a scenario where an Error needs to be raised and to concatenate all of the errors into a single message. I'll add a comment to that effect.